### DISCUSSION TITLE
Question about spring data jdbc and kotlin code gen

### DISCUSSION THREAD

**Discussion by ichenhe** [querydsl/querydsl#3036] [Q&A] [answered]
I have successfully run a project with Kotlin + gradle kts + spring data jpa + querydsl 4.4.

However, now I want to replace JPA with `spring-data-jdbc` which is more concise. Meanwhile, v5.0 released and  brings me confusion.

1. How to complete the integration with data-jdbc? There seems to be only jpa relevant documents.
2. What is [kotlin-codegen](https://github.com/querydsl/querydsl/issues/1828), and how to use it? Now I'm using `kapt("com.querydsl:querydsl-apt:4.4.0:jpa")`, Is it any different from the kotlin-codegen?

Any hints will be appreciated.

**Comment by jwgmeligmeyling** [comment]
@F43nd1r could you explain the kotlin-codegen introduced in v5?

**Comment by F43nd1r** [comment, marked as answer]
`querydsl-kotlin-codegen` does exactly one thing: It replaces all java code generation in `querydsl-apt` with kotlin code. It's only used in addition to the apt module:

     kapt("com.querydsl:querydsl-apt:5.0.0:jpa")
     kapt("com.querydsl:querydsl-kotlin-codegen:5.0.0")

You don't *need* this just because your project is using kotlin, but it brings some advantages:
 - no need for java compilation improves build performance (only if you have no other (generated) java sources)
 - lazy QClass fields allow for arbitrarily deep field access without any configuration even with circular entity dependencies
 - nullsafety (this is just a general advantage of kotlin over java, no specific problem case known to me)

I developed this plugin because I had issues with cross-compilation of kotlin and java, which was solved by this as well, but that's mostly anecdotal.

**Reply by F43nd1r** [reply]
Here's an example of a Java QClass and a Kotlin QClass for comparison:
```java
@Generated("com.querydsl.codegen.DefaultEntitySerializer")
public class QVersion extends EntityPathBase<Version> {

    private static final long serialVersionUID = -738061112L;

    private static final PathInits INITS = PathInits.DIRECT2;

    public static final QVersion version = new QVersion("version");

    public final QApp app;

    public final NumberPath<Integer> code = createNumber("code", Integer.class);

    public final NumberPath<Integer> id = createNumber("id", Integer.class);

    public final StringPath mappings = createString("mappings");

    public final StringPath name = createString("name");

    public QVersion(String variable) {
        this(Version.class, forVariable(variable), INITS);
    }

    public QVersion(Path<Version> path) {
        this(path.getType(), path.getMetadata(), PathInits.getFor(path.getMetadata(), INITS));
    }

    public QVersion(PathMetadata metadata) {
        this(metadata, PathInits.getFor(metadata, INITS));
    }

    public QVersion(PathMetadata metadata, PathInits inits) {
        this(Version.class, metadata, inits);
    }

    public QVersion(Class<? extends Version> type, PathMetadata metadata, PathInits inits) {
        super(type, metadata, inits);
        this.app = inits.isInitialized("app") ? new QApp(forProperty("app"), inits.get("app")) : null;
    }

}
```
```kotlin
@Generated("com.querydsl.kotlin.codegen.KotlinEntitySerializer")
public class QVersion : EntityPathBase<Version> {
  public val app: QApp by lazy {
    QApp(forProperty("app"))
  }


  public val code: NumberPath<Int> = createNumber("code", Int::class.java)

  public val id: NumberPath<Int> = createNumber("id", Int::class.java)

  public val mappings: StringPath = createString("mappings")

  public val name: StringPath = createString("name")

  public constructor(variable: String) : super(Version::class.java, forVariable(variable))

  public constructor(path: Path<Version>) : super(path.type, path.metadata)

  public constructor(metadata: PathMetadata) : super(Version::class.java, metadata)

  public constructor(type: Class<out Version>, metadata: PathMetadata) : super(type, metadata)

  public companion object {
    private const val serialVersionUID: Long = -738061112

    public val version: QVersion = QVersion("version")
  }
}
```
As you can see, in java we have `PathInits.DIRECT2` by default, which limits our deep field access to two levels. In kotlin we don't use `PathInits` at all. Instead entity fields are lazy, which means we don't need to worry about how deep our field access goes.

**Reply by F43nd1r** [reply]
@jwgmeligmeyling should we maybe document this somewhere? And if so, where would you put it?

**Reply by ichenhe** [reply]
Great explanation, thank you. I think it should be documented in case other people are confused.

**Reply by jwgmeligmeyling** [reply]
I think the best part to store this documentation is in the ["Querydsl Reference Guide"](https://querydsl.com/static/querydsl/5.0.0/reference/html_single/#codegen), specificially as a subchapter under "3.3. Code generation".

That page is maintained under https://github.com/querydsl/querydsl/blob/master/querydsl-docs/src/main/docbook/en-US/content/general/codegen.xml .

The documentation is currently written in DocBook, so there is yet another legacy hurdle to overcome before even starting to update the documentation. I am planning to rewrite major portions of the reference guide at the end of the year anyway, as portions of it have become outdated. So I am open to switching to Markdown or Asciidoc as well.

**Reply by ichenhe** [reply]
In my opinion, Markdown is great. It's simple, but it's good enough, and many document generators are compatible with it.

**Comment by alvistar** [comment]
Hi,
I am writing here before raising this as an issue.

I have a project running with Spring Boot 2.5.5 and Spring Data. Gradle version 7.2

In build.kts:
implementation("com.querydsl:querydsl-jpa:5.0.0")
implementation("com.querydsl:querydsl-core:5.0.0")
kapt("com.querydsl:querydsl-apt:5.0.0:jpa")

Everything works, but if I add:

kapt("com.querydsl:querydsl-kotlin-codegen:5.0.0")

Spring will throw this exception:
Caused by: java.lang.IllegalStateException: Could not access method or field: class org.springframework.util.ReflectionUtils cannot access a member of class manta_backend.entities.QAccount with modifiers "private static final"

Any help? I cannot understand if this is more an issue to be raise on Spring Data.

Thanks in advance,
Alessandro

**Reply by F43nd1r** [reply]
querydsl-kotlin-codegen is currently incompatible with spring data. Fix is in https://github.com/querydsl/querydsl/pull/3076

**Reply by alvistar** [reply]
"jwgmeligmeyling closed this 3 hours ago".

Realtime!

Thanks

**Reply by ArchieLabs** [reply]
is this fix pushed into 5.0.0 ? I'm having the same issue and I cannot use QueryDSL in a kotlin project.

` Could not access method or field: class org.springframework.util.ReflectionUtils cannot access a member of class com.test.entities.QAccount with modifiers "private static final"`

**Comment by xmlking** [comment]
still seams like issue with Spring Boot 3.0.2 as stated by @alvistar  above

**Reply by GourdErwa** [reply]
still seams like issue with Spring Boot 3.1.1 as stated by @alvistar above

```java
Caused by: java.lang.IllegalStateException: Could not access method or field: class org.springframework.util.ReflectionUtils cannot access a member of class io.norn.examples.mongodb.customer.QCustomer with modifiers "private static final"
	at org.springframework.util.ReflectionUtils.handleReflectionException(ReflectionUtils.java:107) ~[spring-core-6.0.10.jar:6.0.10]
	at org.springframework.util.ReflectionUtils.getField(ReflectionUtils.java:658) ~[spring-core-6.0.10.jar:6.0.10]
	at org.springframework.data.querydsl.SimpleEntityPathResolver.lambda$createPath$0(SimpleEntityPathResolver.java:74) ~[spring-data-commons-3.1.1.jar:3.1.1]
```
